### ISSUE TITLE
work with hypopg coredump

### ISSUE THREAD

**Issue by HUST-Huyajun** [ossc-db/pg_hint_plan#205]
```
create extension hypopg ;
load 'pg_hint_plan';

create table t1(a int, b int, c int);
create index ON t1 (a);

select hypopg_create_index('create index on t1(b)');
explain select/*+ indexscan(t1 t1_a_idx)*/ from t1 where a = 3 and b =4 ;
-- coredump
```
![image](https://github.com/user-attachments/assets/6e8b7c59-0958-4972-8ad5-fb4d3d3d03a8)
it seems like pg_hint_plan does not consider the impact of hypothetical indexes, restrict_indexes should filter out the hypothetical IndexOptInfo in RelOptInfo

**Comment by michaelpq** [comment]
I'll look at producing a patch about that. For the time being, let's give a poke at @rjuju for the interactions with hypopg.

**Comment by rjuju** [comment]
Unfortunately there is nothing that hypopg can do to avoid that.  It's a simple EXPLAIN so it does add a fake IndexOptInfo with a fake Oid.  Then pg_hint_plan tries to call `get_rel_name()` with that fake index Oid, which is almost guaranteed to not point to a relation.  I agree with @HUST-Huyajun, checking and ignoring hypothetical indexes is likely  the only thing that can be done here.  Unfortunately there is no API that hypopg can implement to let pg_hint_plan get the hypothetical index name, so you can't really use such an index in a hint.

**Comment by michaelpq** [comment]
> I agree with @HUST-Huyajun, checking and ignoring hypothetical indexes is likely the only thing that can be done here.
Yep. I'm on the same line as you here. Ignoring them sounds like a sane option, even if from the user perspective it could be nice to create an hypothetical index and manipulate the plan paths with hints.

**Comment by rjuju** [comment]
Actually, I just remembered that there is an existing hook to get the hypothetical index name, as EXPLAIN needs it:  `explain_get_index_name_hook()`.  The naming is quite explicit, but I don't see any reason why we couldn't actually call the for a hypothetical index.

**Comment by michaelpq** [comment]
The trick with explain_get_index_name_hook() does not seem to address all that properly: we still have problems with EXPLAIN plans where hypothetical indexes have names that do not show up correctly.

It does not seem like I can upload a patch to this ticket (it keeps failing), but I have pushed a patch to a temporary branch in my own fork of the project based on PG17:
https://github.com/michaelpq/pg_hint_plan/tree/hypothetical_index_17

**Comment by michaelpq** [comment]
If there's still a use case for this bug, let's reopen but I don't really see a solution in the short to middle-term.

**Comment by patrobinson** [comment]
Hi @michaelpq , I understand that using hypothetical indexes with index hinting is a bit of a weird interaction between two postgres plugins that can cause unexpected behaviour, but I think we should at the very least not crash the database.

**Comment by michaelpq** [comment]
> Hi @michaelpq , I understand that using hypothetical indexes with index hinting is a bit of a weird interaction between two postgres plugins that can cause unexpected behaviour, but I think we should at the very least not crash the database.

I have not looked at this problem again in months. If you have a proposal of patch, please feel free.

**Comment by patrobinson** [comment]
@michaelpq I've pushed https://github.com/ossc-db/pg_hint_plan/pull/247 to handle any NULL index name
